dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java

```diff
@@ -77,13 +82,87 @@ public boolean isLearnable() {
   }
 
   /**
+   * Transform the given image to column form to facilitate matrix multiplication.
+   * @param imageIndex the index of the image in the input matrix.
+   * @param input input values for this layer.
+   * @return the converted column.
+   */
+  private Matrix im2col(final int imageIndex, final Matrix input) {
+    final int[] inputShape = getInputShape();
+    final Matrix col =
+        matrixFactory.zeros(NeuralNetworkUtils.getShapeLength(outputShape), kernelHeight * kernelWidth);
+    for (int kh = 0; kh < kernelHeight; ++kh) {
+      for (int kw = 0; kw < kernelWidth; ++kw) {
+        int ih = kh - paddingHeight;
+        for (int oh = 0; oh < outputShape[0]; ++oh) {
+          if (ih >= 0 && ih < inputShape[0]) {
+            int iw = kw - paddingWidth;
+            for (int ow = 0; ow < outputShape[1]; ++ow) {
+              if (iw >= 0 && iw < inputShape[1]) {
+                col.put(oh * outputShape[1] + ow,
+                    kh * kernelHeight + kw, input.get(ih * inputShape[1] + iw, imageIndex));
+              }
+              iw += strideWidth;
+            }
+          }
+          ih += strideHeight;
+        }
+      }
+    }
+    return col;
+  }
+
+  /**
+   * Transform the given column form to the image to facilitate matrix multiplication.
+   * @param col the given column.
+   * @return the converted image.
+   */
+  private Matrix col2im(final Matrix col) {
+    final int[] inputShape = getInputShape();
+    final int kernelSize = kernelHeight * kernelWidth;
+    final Matrix im = matrixFactory.zeros(NeuralNetworkUtils.getShapeLength(inputShape), 1);
+    int colIndex = 0;
+    for (int kh = 0; kh < kernelHeight; ++kh) {
+      for (int kw = 0; kw < kernelWidth; ++kw) {
+        int ih = kh - paddingHeight;
+        for (int oh = 0; oh < outputShape[0]; ++oh) {
+          if (ih < 0 || ih >= inputShape[0]) {
+            colIndex += outputShape[1];
+          } else {
+            int iw = kw - paddingWidth;
+            for (int ow = 0; ow < outputShape[1]; ++ow) {
+              if (iw >= 0 && iw < inputShape[1]) {
+                final int colh = colIndex / kernelSize;
+                final int colw = colIndex - colh * kernelSize;
+                final int inputIndex = ih * inputShape[1] + iw;
+                final float newValue = col.get(colh, colw) + im.get(inputIndex, 0);
+                im.put(inputIndex, 0, newValue);
```

Comment:
Why don't we use `im.put(inputIndex, newValue)` instead of `im.put(inputIndex, 0, newValue)`?